File: images/models/connector.py
"""Pydantic models for connector schema — external system bindings."""

import re
from urllib.parse import urlsplit
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator
from typing import Optional, Literal, Union

_INTERVAL_PATTERN = re.compile(r"^\d+[smhd]$")


class ConnectorFollowUp(BaseModel):
    """Follow-up poll config: fetch a per-item URL to get nested items (e.g. Slack thread replies)."""

    model_config = ConfigDict(extra="forbid")

    url: str  # URL template; {field} substituted from parent item, ${ENV} from env
    when: Optional[str] = None  # only follow up when this field is truthy/non-zero on the parent item
    response_key: Optional[str] = None  # extract list from follow-up response (e.g. $.messages)
    dedup_key: Optional[str] = None  # field to use for dedup instead of full-object hash
    skip_first: bool = False  # skip first result (e.g. thread parent repeated in Slack replies)


class ConnectorWebhookAuth(BaseModel):
    """HMAC-SHA256 signature verification for webhook sources (e.g. Slack Events API)."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["hmac_sha256"] = "hmac_sha256"
    secret_env: Optional[str] = None  # env var name containing the signing secret
    secret_credref: Optional[str] = None  # credref name for the signing secret
    header: str = "X-Slack-Signature"  # header carrying the computed signature
    timestamp_header: Optional[str] = "X-Slack-Request-Timestamp"  # for replay attack protection
    prefix: str = "v0="  # prefix on the signature value (Slack uses "v0=")
    challenge_field: Optional[str] = "challenge"  # field name for URL verification handshake (Slack)
    max_skew_seconds: int = 300

    @model_validator(mode="after")
    def validate_secret_source(self) -> "ConnectorWebhookAuth":
        if not self.secret_env and not self.secret_credref:
            raise ValueError("webhook_auth requires either secret_env or secret_credref")
        return self


class ConnectorSource(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["webhook", "poll", "schedule", "channel-watch", "none"]
    payload_schema: Optional[dict] = Field(default=None, alias="schema")
    webhook_auth: Optional[ConnectorWebhookAuth] = None  # HMAC auth for webhook sources
    path: Optional[str] = None
    body_format: Optional[Literal["json", "form_urlencoded", "form_urlencoded_payload", "form_urlencoded_payload_json_field"]] = None
    payload_field: Optional[str] = None
    response_status: Optional[int] = None
    response_body: Optional[str] = None
    response_content_type: Optional[str] = None
    ack_strategy: Optional[str] = None
    # poll fields
    url: Optional[str] = None
    method: str = "GET"
    headers: Optional[dict[str, str]] = None
    interval: Optional[str] = None
    response_key: Optional[str] = None
    dedup_key: Optional[str] = None  # field to use for deduplication instead of full-object hash
    follow_up: Optional[ConnectorFollowUp] = None  # per-item follow-up fetch (e.g. thread replies)
    # schedule fields
    cron: Optional[str] = None
    # channel-watch fields
    channel: Optional[str] = None
    pattern: Optional[str] = None
    # poll extended fields
    transform: Optional[str] = None  # dot-path extraction applied after response_key (e.g. $.data.results)
    auth: Optional[str] = None  # named service grant for authenticated endpoints

    @model_validator(mode="after")
    def validate_source_fields(self) -> "ConnectorSource":
        if self.type == "none":
            inbound_fields = [
                self.webhook_auth,
                self.path,
                self.body_format,
                self.payload_field,
                self.response_status,
                self.response_body,
                self.response_content_type,
                self.ack_strategy,
                self.url,
                self.interval,
                self.response_key,
                self.cron,
                self.channel,
                self.pattern,
                self.transform,
                self.auth,
            ]
            if self.headers is not None:
                inbound_fields.append("set")
            if self.method != "GET":
                inbound_fields.append("set")
            if any(f is not None for f in inbound_fields):
                raise ValueError("source type none does not accept inbound source fields")
        elif self.type == "poll":
            if not self.url:
                raise ValueError("poll source requires 'url'")
            if self.interval and self.cron:
                raise ValueError("poll source: 'interval' and 'cron' are mutually exclusive")
            if not self.interval and not self.cron:
                raise ValueError("poll source requires exactly one of 'interval' or 'cron'")
            if self.interval and not _INTERVAL_PATTERN.match(self.interval):
                raise ValueError(f"Invalid interval format: {self.interval} (expected e.g. '30s', '5m', '1h', '1d')")
        elif self.type == "schedule":
            if not self.cron:
                raise ValueError("schedule source requires 'cron'")
        elif self.type == "channel-watch":
            if not self.channel:
                raise ValueError("channel-watch source requires 'channel'")
            if not self.pattern:
                raise ValueError("channel-watch source requires 'pattern'")
        elif self.type == "webhook":
            if self.path:
                parts = urlsplit(self.path)
                if not self.path.startswith("/") or parts.scheme or parts.netloc or parts.query or parts.fragment:
                    raise ValueError("webhook source path must be an absolute path without query or fragment")
            if self.payload_field and self.body_format not in {"form_urlencoded_payload", "form_urlencoded_payload_json_field"}:
                raise ValueError("payload_field is only valid with body_format 'form_urlencoded_payload_json_field'")
            if self.response_status is not None and not 200 <= self.response_status <= 299:
                raise ValueError("webhook response_status must be a 2xx status code")
            poll_fields = [self.url, self.interval, self.response_key, self.cron, self.channel, self.pattern, self.transform, self.auth]
            if self.headers is not None:
                poll_fields.append("set")
            if self.method != "GET":
                poll_fields.append("set")
            if any(f is not None for f in poll_fields):
                raise ValueError("webhook source does not accept poll/schedule/channel-watch fields")
        elif self.path or self.body_format or self.payload_field or self.response_status is not None or self.response_body is not None or self.response_content_type is not None or self.ack_strategy:
            raise ValueError(f"{self.type} source does not accept webhook body/path fields")
        return self


class ConnectorRelayTarget(BaseModel):
    """Direct HTTP relay: send matched payload to an external endpoint without spawning an agent."""

    model_config = ConfigDict(extra="forbid")

    url: str  # ${ENV} expanded; target endpoint
    method: str = "POST"
    headers: Optional[dict[str, str]] = None  # ${ENV} expanded
    body: str  # Jinja2 template rendered with payload fields, then ${ENV} expanded
    content_type: str = "application/json"


class ConnectorRoute(BaseModel):
    model_config = ConfigDict(extra="forbid")

    match: dict[str, Union[str, list[str], None]]
    # Agent/team routing: deliver a task to an agent or team via DM channel
    target: Optional[dict[str, str]] = None
    # Relay routing: POST directly to an HTTP endpoint, no agent spawned
    relay: Optional[ConnectorRelayTarget] = None
    priority: Literal["high", "normal", "low"] = "normal"
    sla: Optional[str] = None
    brief: Optional[str] = None  # Jinja2 template for task brief delivered to agent

    @model_validator(mode="after")
    def validate_routing(self) -> "ConnectorRoute":
        has_target = self.target is not None
        has_relay = self.relay is not None
        if not has_target and not has_relay:
            raise ValueError("Route must specify either 'target' (agent/team) or 'relay' (HTTP endpoint)")
        if has_target and has_relay:
            raise ValueError("Route cannot specify both 'target' and 'relay'")
        return self


class ConnectorMCPTool(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    method: str = "GET"
    path: Optional[str] = None
    parameters: Optional[dict] = None
    input_schema: Optional[dict] = None
    returns: Optional[dict] = None
    description: str = ""
    requires_config: Optional[str] = None
    query_params: list[str] = Field(default_factory=list)
    whitelist_check: Optional[str] = None
    requires_consent_token: Optional[dict] = None

    @model_validator(mode="after")
    def validate_tool_controls(self) -> "ConnectorMCPTool":
        if not self.path and not self.input_schema:
            raise ValueError("tool requires either path or input_schema")
        params = set((self.parameters or self.input_schema or {}).keys())
        if self.whitelist_check and self.whitelist_check not in params:
            raise ValueError(f"whitelist_check references unknown parameter {self.whitelist_check!r}")
        for field in self.query_params:
            if field not in params:
                raise ValueError(f"query_params references unknown parameter {field!r}")
        if self.requires_consent_token:
            operation_kind = self.requires_consent_token.get("operation_kind")
            token_field = self.requires_consent_token.get("token_input_field")
            target_field = self.requires_consent_token.get("target_input_field")
            if not operation_kind:
                raise ValueError("requires_consent_token.operation_kind is required")
            if not token_field:
                raise ValueError("requires_consent_token.token_input_field is required")
            if not target_field:
                raise ValueError("requires_consent_token.target_input_field is required")
            if token_field not in params:
                raise ValueError(f"requires_consent_token references unknown token_input_field {token_field!r}")
            if target_field not in params:
                raise ValueError(f"requires_consent_token references unknown target_input_field {target_field!r}")
        return self


class ConnectorMCP(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    credential: str
    api_base: Optional[str] = None
    server: Optional[str] = None
    tools: Optional[list[ConnectorMCPTool]] = None


class ConnectorCredential(BaseModel):
    """A single credential requirement for a connector."""
    model_config = ConfigDict(extra="forbid")
    name: str
    description: str = ""
    type: Literal["secret", "config"] = "secret"
    scope: Literal["service-grant", "env-var", "file"] = "service-grant"
    grant_name: Optional[str] = None
    setup_url: Optional[str] = None
    example: Optional[str] = None


class ConnectorAuth(BaseModel):
    """Authentication method for a connector's external API."""
    model_config = ConfigDict(extra="forbid")
    type: Literal["none", "bearer", "jwt-exchange", "oauth2", "google_service_account"] = "none"
    token_url: Optional[str] = None
    token_params: Optional[dict[str, str]] = None
    token_response_field: str = "access_token"
    token_ttl_seconds: int = 3600
    scopes: list[str] = Field(default_factory=list)


class ConnectorRequires(BaseModel):
    model_config = ConfigDict(extra="forbid")

    services: list[str] = Field(default_factory=list)
    credentials: list[ConnectorCredential] = Field(default_factory=list)
    auth: Optional[ConnectorAuth] = None
    egress_domains: list[str] = Field(default_factory=list)


class ConnectorRateLimits(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_per_hour: int = 100
    max_concurrent: int = 10


class GraphIngestNode(BaseModel):
    """Node upsert rule for graph_ingest."""
    kind: str
    label: str                         # Jinja2 template
    properties: dict[str, str] = {}    # Jinja2 templates for values


class GraphIngestEdge(BaseModel):
    """Edge upsert rule for graph_ingest."""
    relation: str
    from_label: str                    # Jinja2 template
    to_kind: str
    to_label: str                      # Jinja2 template


class CorrelateConfig(BaseModel):
    """Cross-source correlation config for graph_ingest rules."""
    source: str              # Name of another active connector
    on: str                  # Field name to join on
    window_seconds: int = 60


class GraphIngestRule(BaseModel):
    """Single graph_ingest rule with optional match filter."""
    match: Optional[dict] = None       # Same semantics as route match; None = all events
    nodes: list[GraphIngestNode] = []
    edges: list[GraphIngestEdge] = []
    correlate: Optional[CorrelateConfig] = None


class ConnectorConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["connector"] = "connector"
    name: str
    version: str = "1.0.0"
    description: str = ""
    author: str = ""
    license: str = ""
    requires: Optional[ConnectorRequires] = None
    source: ConnectorSource
    routes: list[ConnectorRoute] = []
    mcp: Optional[ConnectorMCP] = None
    config: dict = Field(default_factory=dict)
    tools: list[ConnectorMCPTool] = []
    runtime: Optional[dict] = None
    rate_limits: ConnectorRateLimits = Field(default_factory=ConnectorRateLimits)
    graph_ingest: list[GraphIngestRule] = []

    @model_validator(mode="after")
    def _routes_or_graph_ingest(self) -> "ConnectorConfig":
        if self.source.type == "none":
            if self.routes:
                raise ValueError("tool-only connectors must not define routes")
            if not self.tools and not (self.mcp and self.mcp.tools):
                raise ValueError("tool-only connectors must define at least one tool")
            return self
        if not self.routes and not self.graph_ingest and self.mcp is None:
            raise ValueError("Connector must define at least one route, graph_ingest rule, or MCP tool")
        return self

File: internal/models/connector.go
// agency-gateway/internal/models/connector.go
package models

import (
	"fmt"
	"regexp"
	"strings"
)

var intervalPattern = regexp.MustCompile(`^\d+[smhd]$`)

// ConnectorFollowUp defines a follow-up request after an initial poll.
type ConnectorFollowUp struct {
	URL         string  `yaml:"url" validate:"required"`
	When        *string `yaml:"when,omitempty"`
	ResponseKey *string `yaml:"response_key,omitempty"`
	DedupKey    *string `yaml:"dedup_key,omitempty"`
	SkipFirst   bool    `yaml:"skip_first,omitempty"`
}

// ConnectorWebhookAuth defines HMAC-based webhook authentication.
type ConnectorWebhookAuth struct {
	Type            string  `yaml:"type,omitempty" default:"hmac_sha256"`
	SecretEnv       string  `yaml:"secret_env,omitempty"`
	SecretCredref   string  `yaml:"secret_credref,omitempty"`
	Header          string  `yaml:"header,omitempty" default:"X-Slack-Signature"`
	TimestampHeader *string `yaml:"timestamp_header,omitempty"`
	Prefix          string  `yaml:"prefix,omitempty" default:"v0="`
	ChallengeField  *string `yaml:"challenge_field,omitempty"`
	MaxSkewSeconds  int     `yaml:"max_skew_seconds,omitempty" default:"300"`
}

// ConnectorSource defines the inbound event source for a connector.
type ConnectorSource struct {
	Type                string                 `yaml:"type" validate:"required,oneof=webhook poll schedule channel-watch none"`
	PayloadSchema       map[string]interface{} `yaml:"schema,omitempty"`
	WebhookAuth         *ConnectorWebhookAuth  `yaml:"webhook_auth,omitempty"`
	Path                *string                `yaml:"path,omitempty"`
	BodyFormat          *string                `yaml:"body_format,omitempty"`
	PayloadField        *string                `yaml:"payload_field,omitempty"`
	ResponseStatus      *int                   `yaml:"response_status,omitempty"`
	ResponseBody        *string                `yaml:"response_body,omitempty"`
	ResponseContentType *string                `yaml:"response_content_type,omitempty"`
	AckStrategy         *string                `yaml:"ack_strategy,omitempty"`
	URL                 *string                `yaml:"url,omitempty"`
	Method              string                 `yaml:"method,omitempty" default:"GET"`
	Headers             map[string]string      `yaml:"headers,omitempty"`
	Interval            *string                `yaml:"interval,omitempty"`
	ResponseKey         *string                `yaml:"response_key,omitempty"`
	DedupKey            *string                `yaml:"dedup_key,omitempty"`
	FollowUp            *ConnectorFollowUp     `yaml:"follow_up,omitempty"`
	Cron                *string                `yaml:"cron,omitempty"`
	Channel             *string                `yaml:"channel,omitempty"`
	Pattern             *string                `yaml:"pattern,omitempty"`
}

// Validate implements cross-field validation for ConnectorSource.
func (cs *ConnectorSource) Validate() error {
	if cs.BodyFormat != nil {
		switch *cs.BodyFormat {
		case "json", "form_urlencoded", "form_urlencoded_payload", "form_urlencoded_payload_json_field":
		default:
			return fmt.Errorf("unsupported body_format: %s", *cs.BodyFormat)
		}
	}

	switch cs.Type {
	case "none":
		hasInboundFields := cs.WebhookAuth != nil ||
			cs.Path != nil ||
			cs.BodyFormat != nil ||
			cs.PayloadField != nil ||
			cs.ResponseStatus != nil ||
			cs.ResponseBody != nil ||
			cs.ResponseContentType != nil ||
			cs.AckStrategy != nil ||
			cs.URL != nil ||
			cs.Interval != nil ||
			cs.ResponseKey != nil ||
			cs.Cron != nil ||
			cs.Channel != nil ||
			cs.Pattern != nil ||
			cs.Headers != nil ||
			cs.FollowUp != nil ||
			(cs.Method != "" && cs.Method != "GET")
		if hasInboundFields {
			return fmt.Errorf("source type none does not accept inbound source fields")
		}
	case "poll":
		if cs.URL == nil || *cs.URL == "" {
			return fmt.Errorf("poll source requires 'url'")
		}
		if cs.Interval == nil || *cs.Interval == "" {
			return fmt.Errorf("poll source requires 'interval'")
		}
		if !intervalPattern.MatchString(*cs.Interval) {
			return fmt.Errorf("Invalid interval format: %s", *cs.Interval)
		}
	case "schedule":
		if cs.Cron == nil || *cs.Cron == "" {
			return fmt.Errorf("schedule source requires 'cron'")
		}
	case "channel-watch":
		if cs.Channel == nil || *cs.Channel == "" {
			return fmt.Errorf("channel-watch source requires 'channel'")
		}
		if cs.Pattern == nil || *cs.Pattern == "" {
			return fmt.Errorf("channel-watch source requires 'pattern'")
		}
	case "webhook":
		if cs.Path != nil && (*cs.Path == "" || !strings.HasPrefix(*cs.Path, "/")) {
			return fmt.Errorf("webhook source path must start with '/'")
		}
		if cs.ResponseStatus != nil && (*cs.ResponseStatus < 200 || *cs.ResponseStatus > 299) {
			return fmt.Errorf("webhook response_status must be a 2xx status code")
		}
		if cs.PayloadField != nil {
			if cs.BodyFormat == nil || (*cs.BodyFormat != "form_urlencoded_payload" && *cs.BodyFormat != "form_urlencoded_payload_json_field") {
				return fmt.Errorf("payload_field is only valid with body_format 'form_urlencoded_payload_json_field'")
			}
		}
		hasPollFields := cs.URL != nil ||
			cs.Interval != nil ||
			cs.ResponseKey != nil ||
			cs.Cron != nil ||
			cs.Channel != nil ||
			cs.Pattern != nil ||
			cs.Headers != nil ||
			(cs.Method != "" && cs.Method != "GET")
		if hasPollFields {
			return fmt.Errorf("webhook source does not accept poll/schedule/channel-watch fields")
		}
	default:
		if cs.Path != nil || cs.BodyFormat != nil || cs.PayloadField != nil || cs.ResponseStatus != nil || cs.ResponseBody != nil || cs.ResponseContentType != nil || cs.AckStrategy != nil {
			return fmt.Errorf("%s source does not accept webhook body/path fields", cs.Type)
		}
	}
	if cs.WebhookAuth != nil && cs.WebhookAuth.SecretEnv == "" && cs.WebhookAuth.SecretCredref == "" {
		return fmt.Errorf("webhook_auth requires either secret_env or secret_credref")
	}
	return nil
}

// ConnectorRelayTarget defines an HTTP relay destination.
type ConnectorRelayTarget struct {
	URL         string            `yaml:"url" validate:"required"`
	Method      string            `yaml:"method,omitempty" default:"POST"`
	Headers     map[string]string `yaml:"headers,omitempty"`
	Body        string            `yaml:"body" validate:"required"`
	ContentType string            `yaml:"content_type,omitempty" default:"application/json"`
}

// ConnectorRoute defines a routing rule for matched events.
type ConnectorRoute struct {
	Match        map[string]interface{} `yaml:"match" validate:"required"`
	Target       map[string]string      `yaml:"target,omitempty"`
	Relay        *ConnectorRelayTarget  `yaml:"relay,omitempty"`
	HandlingMode string                 `yaml:"handling_mode,omitempty" validate:"omitempty,oneof=async_ack sync_response"`
	Priority     string                 `yaml:"priority,omitempty" validate:"omitempty,oneof=high normal low" default:"normal"`
	SLA          *string                `yaml:"sla,omitempty"`
	Brief        *string                `yaml:"brief,omitempty"`
}

// Validate implements target/relay mutual exclusion for ConnectorRoute.
func (cr *ConnectorRoute) Validate() error {
	hasTarget := cr.Target != nil
	hasRelay := cr.Relay != nil

	if !hasTarget && !hasRelay {
		return fmt.Errorf("Route must specify either 'target' or 'relay'")
	}
	if hasTarget && hasRelay {
		return fmt.Errorf("Route cannot specify both 'target' and 'relay'")
	}
	return nil
}

// ConnectorMCPTool defines a single MCP tool exposed by a connector.
type ConnectorMCPTool struct {
	Name                 string                 `yaml:"name" validate:"required"`
	Method               string                 `yaml:"method" default:"GET"`
	Path                 string                 `yaml:"path"`
	Parameters           map[string]interface{} `yaml:"parameters"`
	InputSchema          map[string]interface{} `yaml:"input_schema"`
	Returns              map[string]interface{} `yaml:"returns"`
	Description          string                 `yaml:"description"`
	RequiresConfig       string                 `yaml:"requires_config,omitempty"`
	QueryParams          []string               `yaml:"query_params,omitempty"`
	WhitelistCheck       string                 `yaml:"whitelist_check,omitempty"`
	RequiresConsentToken *ConsentRequirement    `yaml:"requires_consent_token,omitempty"`
}

func (ct *ConnectorMCPTool) Validate() error {
	if ct.Path == "" && len(ct.Parameters) == 0 && len(ct.InputSchema) == 0 {
		return fmt.Errorf("tool %q requires path, parameters, or input_schema", ct.Name)
	}
	schema := ct.Parameters
	if len(schema) == 0 {
		schema = ct.InputSchema
	}
	params := make(map[string]bool, len(schema))
	for name := range schema {
		params[name] = true
	}
	if ct.WhitelistCheck != "" && !params[ct.WhitelistCheck] {
		return fmt.Errorf("tool %q whitelist_check references unknown parameter %q", ct.Name, ct.WhitelistCheck)
	}
	for _, field := range ct.QueryParams {
		if !params[field] {
			return fmt.Errorf("tool %q query_params references unknown parameter %q", ct.Name, field)
		}
	}
	return validateConsentRequirement(ct.Name, ct.RequiresConsentToken, params)
}

// ConnectorMCP defines the MCP server configuration for a connector.
type ConnectorMCP struct {
	Name       string             `yaml:"name" validate:"required"`
	Credential string             `yaml:"credential" validate:"required"`
	APIBase    *string            `yaml:"api_base"`
	Server     *string            `yaml:"server"`
	Tools      []ConnectorMCPTool `yaml:"tools"`
}

// ConnectorCredential defines a credential required by a connector.
type ConnectorCredential struct {
	Name        string `yaml:"name" json:"name" validate:"required"`
	Description string `yaml:"description" json:"description"`
	Type        string `yaml:"type" json:"type" default:"secret"`          // secret | config
	Scope       string `yaml:"scope" json:"scope" default:"service-grant"` // service-grant | env-var | file
	GrantName   string `yaml:"grant_name" json:"grant_name,omitempty"`
	SetupURL    string `yaml:"setup_url" json:"setup_url,omitempty"`
	Example     string `yaml:"example" json:"example,omitempty"`
}

// ConnectorAuth defines authentication configuration for a connector.
type ConnectorAuth struct {
	Type               string            `yaml:"type" json:"type" default:"none"` // none | bearer | jwt-exchange | oauth2 | google_service_account
	TokenURL           string            `yaml:"token_url" json:"token_url,omitempty"`
	TokenParams        map[string]string `yaml:"token_params" json:"token_params,omitempty"`
	TokenResponseField string            `yaml:"token_response_field" json:"token_response_field" default:"access_token"`
	TokenTTLSeconds    int               `yaml:"token_ttl_seconds" json:"token_ttl_seconds" default:"3600"`
	Scopes             []string          `yaml:"scopes" json:"scopes,omitempty"`
}

// ConnectorRequires lists service dependencies for a connector.
type ConnectorRequires struct {
	Services      []string              `yaml:"services" json:"services"`
	Credentials   []ConnectorCredential `yaml:"credentials" json:"credentials"`
	Auth          *ConnectorAuth        `yaml:"auth" json:"auth,omitempty"`
	EgressDomains []string              `yaml:"egress_domains" json:"egress_domains"`
}

// ConnectorRateLimits defines rate limiting parameters for a connector.
type ConnectorRateLimits struct {
	MaxPerHour    int `yaml:"max_per_hour" default:"100"`
	MaxConcurrent int `yaml:"max_concurrent" default:"10"`
}

// ConnectorConfig is the schema for connector YAML files.
type ConnectorConfig struct {
	Kind        string                 `yaml:"kind" validate:"required,oneof=connector" default:"connector"`
	Name        string                 `yaml:"name" validate:"required"`
	Version     string                 `yaml:"version" default:"1.0.0"`
	Description string                 `yaml:"description"`
	Author      string                 `yaml:"author"`
	License     string                 `yaml:"license,omitempty"`
	Requires    *ConnectorRequires     `yaml:"requires"`
	Source      ConnectorSource        `yaml:"source" validate:"required"`
	Config      map[string]interface{} `yaml:"config,omitempty"`
	Routes      []ConnectorRoute       `yaml:"routes"`
	MCP         *ConnectorMCP          `yaml:"mcp"`
	Tools       []ConnectorMCPTool     `yaml:"tools,omitempty"`
	Runtime     map[string]interface{} `yaml:"runtime,omitempty"`
	RateLimits  ConnectorRateLimits    `yaml:"rate_limits"`
}

// Validate implements cross-field validation for ConnectorConfig.
func (cc *ConnectorConfig) Validate() error {
	if err := cc.Source.Validate(); err != nil {
		return err
	}
	if len(cc.Routes) == 0 && cc.Source.Type != "none" {
		return fmt.Errorf("Connector must define at least one route")
	}
	if cc.Source.Type == "none" && len(cc.Routes) != 0 {
		return fmt.Errorf("tool-only connectors must not define routes")
	}
	if len(cc.Routes) == 0 && cc.Source.Type == "none" && cc.MCP == nil && len(cc.Tools) == 0 {
		return fmt.Errorf("tool-only connectors must define at least one tool")
	}

	for i := range cc.Routes {
		if err := cc.Routes[i].Validate(); err != nil {
			return err
		}
	}
	if cc.MCP != nil {
		for i := range cc.MCP.Tools {
			if err := cc.MCP.Tools[i].Validate(); err != nil {
				return err
			}
		}
	}
	for i := range cc.Tools {
		if err := cc.Tools[i].Validate(); err != nil {
			return err
		}
	}

	return nil
}
